### PR TITLE
drivers: serial: nrfx_uarte: Add low power mode

### PR DESCRIPTION
**Pull request by nordic-krch**
Lowest power consumption can be achieved when uarte peripheral
is disabled when not used. In low power mode, need for both
directions is tracked and if both are no in use peripheral is
disabled. TX disabling is instant but RX requires flushing RX
fifo because data in hardware fifo is lost when peripheral is
re-enabled.

In case of interrupt driver API, low power mode can only be used when RX is not used because otherwise receiver is always on. In case of async API, receiver can be disabled thus need for TX and RX is tracked and UARTE is disabled when both directions are not in use.

PR consist of initial driver refactoring. It has one functional change: RX_DISABLED event is generated from RXTO interrupt instead of ENDRX. Previously, rx was not stopped when there was no buffers and receiver was disabled.

### CODE REVIEW

**Review by dcpleung — APPROVED** [review]

**Comment by ioannisg** [comment]
@anangl @Mierunski pls, take a look

**Comment by nordic-krch** [comment]
@anangl @Mierunski ping

File: drivers/serial/uart_nrfx_uarte.c [hunk 1]
```diff
@@ -658,13 +779,44 @@ static int uarte_nrfx_rx_enable(const struct device *dev, uint8_t *buf,
 	data->async->rx_offset = 0;
 	data->async->rx_next_buf = NULL;
 	data->async->rx_next_buf_len = 0;
+
+	if (get_dev_config(dev)->flags &  UARTE_CFG_FLAG_LOW_POWER) {
```
Comment (Mierunski) [inline drivers/serial/uart_nrfx_uarte.c:783]: Unnecesary space

File: drivers/serial/uart_nrfx_uarte.c [hunk 2]
```diff
@@ -963,49 +1081,138 @@ static void endrx_isr(const struct device *dev)
...
+ *
+ * @return number of bytes flushed from the fifo.
+ */
+static uint8_t rx_flush(const struct device *dev, uint8_t *buf, uint32_t len)
+{
+	/* Flushing RX fifo requires buffer bigger than 4 bytes to empty fifo*/
+	static const uint8_t dirty;
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	uint32_t prev_rx_amount = nrf_uarte_rx_amount_get(uarte);
+	uint8_t tmp_buf[5];
```
Comment (Mierunski) [inline drivers/serial/uart_nrfx_uarte.c:1121]: Same magic number

**Review by Mierunski — CHANGES_REQUESTED** [review]

**Comment by nordic-krch** [comment]
@Mierunski thanks for review, comments applied.

**Review by Mierunski — APPROVED** [review]

**Pull request merged**